paddle/fluid/framework/new_executor/new_executor_defs.h

```diff
@@ -488,9 +490,24 @@ class VariableScope : public ScopeBase {
     info.var_ref_count_ = 0;
     info.vardesc_ = nullptr;
     vec_meta_info_.push_back(info);
-    scope_ptr_.reset(new Scope());
+    outer_scope_ = nullptr;
   }
-  const Scope* GetScope() const { return scope_ptr_.get(); }
+
+  ~VariableScope() {
+    outer_scope_->DelListener(this);  // don't listen anymore.
+  }
+
+  void SetScope(Scope* outer_scope) {
+    // We don't consider SetScope twice currently.
+    PADDLE_ENFORCE_EQ(
+        outer_scope_, nullptr,
+        platform::errors::PreconditionNotMet(
+            "You have been SetScope(), please don't call SetScope twice."));
+    outer_scope_ = outer_scope;
+    outer_scope->AddListener(this);
+  }
```

Comment:
U can directly initialize outer_scope_ in constructor, and do not expose a `SetScope` method.